### PR TITLE
identity: Don't let subs be re-registered

### PR DESCRIPTION
**Pull request by gavofyork**
Also allow subs to de-register themselves and collect the deposit.

Also allow individual registering and removal of subs.

Also remove some double-spaces.

TODO:
- [x] Tests
- [x] Benchmarks
- [x] Weights

### CODE REVIEW

File: frame/identity/src/lib.rs [hunk 1]
```diff
@@ -1159,6 +1198,91 @@ decl_module! {
 				id.info.additional.len() as Weight // X
 			)).into())
 		}
+
+		/// Remove the given account from the sender's subs.
```
Comment (joepetrowski) [inline frame/identity/src/lib.rs:1202]:
```suggestion
		/// Add the given account to the sender's subs.
```

File: frame/identity/src/lib.rs [hunk 2]
```diff
@@ -1159,6 +1204,103 @@ decl_module! {
...
+		/// Payment: Balance reserved by a previous `set_subs` call for one sub will be repatriated
+		/// to the sender (*not* the original depositor).
+		///
+		/// The dispatch origin for this call must be _Signed_ and the sender must have a registered
+		/// super-identity.
+		///
+		/// NOTE: This should not normally be used, but is provided in the case that the non-
+		/// controller of an account is maliciously registered as a sub-account.
+		#[weight = weight_for::quit_sub::<T>(
+			T::MaxSubAccounts::get().into(), // S
```
Comment (apopiak) [inline frame/identity/src/lib.rs:1291]: nit: I think we should do weight refund in this case so we only force people to pay for the amount of subs they have.

Comment (gavofyork) [reply frame/identity/src/lib.rs:1291]: well, i don't think it'll be a huge issue. the amount of subs is capped on purpose.

File: frame/identity/src/lib.rs [hunk 3]
```diff
@@ -774,6 +817,9 @@ decl_module! {
 			let (old_deposit, old_ids) = <SubsOf<T>>::get(&sender);
 			let new_deposit = T::SubAccountDeposit::get() * <BalanceOf<T>>::from(subs.len() as u32);
 
+			let not_other_sub = subs.iter().filter_map(|i| SuperOf::<T>::get(&i.0)).all(|i| &i.0 == &sender);
```
[the remark this answers is not in the document]
Comment (gavofyork) [reply frame/identity/src/lib.rs:820]: well, it's out of scope for the current review, but needs to be done, yeah. maybe one of the weights gods can lend a hand?

[the remark this answers is not in the document]
Comment (gavofyork) [reply frame/identity/src/lib.rs:828]: done

**Comment by gavofyork** [comment]
Added some weights generated locally. @shawntabrizi i guess the standard machine weights can go in along with the Weight trait impl.

**Pull request merged**